pkg/telemetryquerylanguage/tql/functions.go

```diff
@@ -21,10 +21,12 @@ import (
 
 type PathExpressionParser func(*Path) (GetSetter, error)
 
+type EnumParser func(*Path) (*Enum, bool)
```

Comment:
What do you think about making EnumParser returning an error rather than just a bool? That gives a place for the implemented EnumParser to be more informative than just "NO". ("'IFNO' not found -- did you mean 'INFO'?" is one example).

Reply:
The tricky part would be trying to differentiate between an enum that didn't match in the symbol table and a Path that needs to be interpreted as part as a Path.

This could be solved if we can make the grammar smart enough to differentiate between a Path and an Enum (as right now they are the same according to the lexer). I started out with that but ended up with this solution bc I didn't like having to add something before, after, or around the Enum to differentiate it from a Path.

Reply:
I probably wouldn't say that it has to have a marker for Enum. Right now, a Path is anything that contains a dot `.`, and an Invocation has to contain parens. So we could, for example, constrain Enum to *not* contain a dot. Personally, I'd prefer saying that it has to be something that looks like an Ident (`[A-Za-z_][A-Za-z0-9_]`), and maybe even specifically an uppercase identifier (`[A-Z_][A-Z0-9_]`). 

What do you think?

Reply:
OK, I read the definition too quickly -- but does it have to be that way? There are two questions here:
- Could we constrain Enum (say) to be an Ident? I'd like to, even if Paths have to be the way they are.
- Could we constrain Paths to be more restrictive? I am probably missing something, but all of the examples I know of contain a `.` -- the grammar could use a `+` instead of `*` and enforce this restriction.

I'm sorry if it feels like I'm being nitpicky here -- I am thinking that in a general-purpose language, having more restrictions on symbols helps in writing grammars that can actually be helpful in their error messages, and if we continue with a "Path is anything" model, then it will likely complicate things later. These sorts of restrictions are easy to remove, but difficult to add.

Reply:
I don't think the path can be more restrictive, thats the issue I ran into when I force implemented this in the transform processor 😞   

Path is the connection between the grammar and the signal's context.  For the transformprocessor it is how we represent the underlying signal's pdata model. So when we want to access a span's name we use the path `name`.  When we want to access a data point's start time you use the path `start_time_unix_nano`. 

There are other aspects of the data model we need to access such as `resource`, `instrumentation_scope` and `metric`. We use the dots to represent these hierarchical relationships, but they are the exception, not the norm.  Most pdata fields are accessed via a path without a dot.

Looking through `traces.go`, `metrics.go`, and `logs.go` I think we **do** represent all pdata paths as lowercase.  We might be able to differentiate between an Enum and a Path if we can enforce all Enums to be UPPERCASE and all paths to be lowercase.  I will look into that.

Reply:
At the moment both Path and Invocation rely on the @Ident token which allows both cases.  I think to enforce Paths being lower case and Enums being upper case we'd need new tokens to enforce that, but they would conflict with @Ident right?  Anything that would match an only UPPERCASE token or only lowercase token would also match the @Ident token.

Reply:
Yes, Ident would have to change. I would like to impose these restrictions, but I won't insist on it.

Reply:
Any ideas on how to change Ident?

Reply:
We might be able to take advantage of this capability in the grammar library:

`"...":<identifier> Match the literal, specifying the exact lexer token type to match.`

Reply:
NVM, it would have to be a literal, which isn't applicable here

Reply:
@kentquirk I think I got it.  I removed Ident and made uppercase and lowercase tokens.  Now the grammar looks like:

- Uppercase for Enum
- Lowercase for Path
- (Uppercase | Lowercase)+ for Invocation.Function

This allows EnumParser to return an error like PathExpressionParser.  All other aspects of the grammar should rename the same.  Path are forced to be lowercase now, but I don't think that is a problem since the transform processor has already established the patter of Paths being lower_snake_case.